### ISSUE TITLE
Thread safety

### ISSUE THREAD

**Issue by jihiggins** [Auburn/FastNoise2#100]
If I call this function from multiple threads at the same time (it's currently using 8 worker threads):

```cpp
#include "FastNoise/FastNoise.h"

void Generate2dNoise(
        int xOffset, int yOffset, int xSize, int ySize, int seed, int octaves, float frequency, float* output
) {
    auto fnSimplex = FastNoise::New<FastNoise::OpenSimplex2>();
    auto fnFractal = FastNoise::New<FastNoise::FractalFBm>();

    fnFractal->SetSource(fnSimplex);
    fnFractal->SetOctaveCount(octaves);
    fnFractal->GenUniformGrid2D(
            output, xOffset, yOffset, xSize, ySize, frequency, seed
    );
}
```

I get random segfaults every once in awhile (it seems to happen pretty early on, or not at all)
Relevant part of the callstack:
![image](https://user-images.githubusercontent.com/9329131/190175603-87cf1f4b-979d-486a-a1fe-445c73866442.png)

Using a global mutex around that function solves the issue, so it seems to be something concurrency related.

(Is there any obvious issue with how I'm using the library / is FastNoise2 actually meant to be threadsafe?)

**Comment by Auburn** [comment]
There was a bug found in the thread safety of SmartNodes, the latest release (0.9.5) contains the fix for it

**Comment by jihiggins** [comment]
> There was a bug found in the thread safety of SmartNodes, the latest release (0.9.5) contains the fix for it

This happens with 0.9.5-alpha (from the release download)

**Comment by Auburn** [comment]
From your code above it looks like you are creating a new SmartNode per function call, ie no cross-thread access for a given node, is that true? or is the code above simplified?

**Comment by jihiggins** [comment]
That's the function being called each time (it's just copy pasted directly.) There's an FFI boundary to get to that function, which I should have mentioned in the original ticket (sorry, just forgot to include it.) No cross thread access for a given node, or at least, there shouldn't be afaict.
The FFI code is basically this right now:
```Rust
    let mut result =
        Vec::with_capacity(x_size as usize * y_size as usize);
    result.resize(result.capacity(), 0.0);
    let _lock = MUTEX.lock();
    unsafe {
        ffi::Generate2dNoise(
            x_offset.into(),
            y_offset.into(),
            x_size.into(),
            y_size.into(),
            seed.into(),
            octaves.into(),
            frequency,
            result.as_mut_ptr(),
        );
    }
    return result;
```

**Comment by jihiggins** [comment]
(If this is only showing up for me, I can try to repro it in just C++ later)

**Comment by Auburn** [comment]
Is your `xSize * ySize < 8` it could be the same bug as #89 I'm working on a fix for it now

**Comment by jihiggins** [comment]
They're all 32x32 (so xSize * ySize = 1024)

I'll see if I can come up with a better set of repro steps, but it also hasn't happened again today. (Maybe I missed clearing out some build artifacts from the old version? Or could just be some weird thing affecting a race condition? It did seem to happen more often when the rest of the app was running at higher framerates, for whatever reason)

**Comment by Auburn** [comment]
Can you see what line of code it is crashing on?

**Comment by jihiggins** [comment]
It's hard to pinpoint exactly right now because it's mangled from the linker / can only get disassembly, but it looks like it's somewhere in this:
[GenUniformGrid2D](https://github.com/Auburn/FastNoise2/blob/v0.9.5-alpha/include/FastNoise/Generators/Generator.inl#L102) or [here](https://github.com/Auburn/FastNoise2/blob/v0.9.5-alpha/include/FastNoise/Generators/Generator.inl#L119)
and then here: (from the assembly it seems like it's when the function returns? which seems weird)
[OpenSimplex2 generator](https://github.com/Auburn/FastNoise2/blob/v0.9.5-alpha/include/FastNoise/Generators/Simplex.inl#L267)

**Comment by Auburn** [comment]
Could you show the assembly where it crashes? Can you get the crash on a debug build for more info?

**Comment by jihiggins** [comment]
Yep! I'll see if I can setup the debug build etc, if I can't get that working I'll try setting up a purely C++ repro using the same compiler tooling etc later (probably won't have time this week for that one tho)

**Comment by jihiggins** [comment]
Here's the disassembly! haven't seen it crash with a debug build yet, though: https://gist.github.com/jihiggins/164ed62dd843e59c736f5ba3b208d1cf#file-disassembly-asm-L48

**Comment by Auburn** [comment]
Thanks. In `FastNoise_Config.h` could you change `#define FASTNOISE_USE_SHARED_PTR false` to `true` will help determine if it's a SmartNode issue or something else

**Comment by jihiggins** [comment]
That's set, got a new crash from an assertion in debug mode.

### First assertion:
https://github.com/Auburn/FastNoise2/blob/v0.9.5-alpha/include/FastNoise/Generators/Generator.h#L141
https://github.com/Auburn/FastNoise2/blob/v0.9.5-alpha/include/FastNoise/Generators/Fractal.h#L10
`    fnFractal->SetSource(fnSimplex);` (from the first post)

### If I continue, next assert is:
https://github.com/Auburn/FastNoise2/blob/v0.9.5-alpha/include/FastNoise/Generators/Generator.inl#L42
https://github.com/Auburn/FastNoise2/blob/v0.9.5-alpha/include/FastNoise/Generators/Generator.h#L143
https://github.com/Auburn/FastNoise2/blob/v0.9.5-alpha/include/FastNoise/Generators/Fractal.h#L10
`    fnFractal->SetSource(fnSimplex);` (from the first post)

### Then:
![image](https://user-images.githubusercontent.com/9329131/192536886-8d114152-dc96-45fd-9f8c-ddfc0ff04e5d.png)
https://github.com/Auburn/FastNoise2/blob/v0.9.5-alpha/include/FastNoise/Generators/Generator.inl#L61
https://github.com/Auburn/FastNoise2/blob/v0.9.5-alpha/include/FastNoise/Generators/Fractal.inl#L24
https://github.com/Auburn/FastNoise2/blob/v0.9.5-alpha/include/FastNoise/Generators/Fractal.inl#L15
https://github.com/Auburn/FastNoise2/blob/v0.9.5-alpha/include/FastNoise/Generators/Generator.inl#L102

### And finally, access violation here:
![image](https://user-images.githubusercontent.com/9329131/192537040-cd24438c-aa6e-40d7-a5eb-26d6a410baac.png)
https://github.com/Auburn/FastNoise2/blob/v0.9.5-alpha/include/FastNoise/Generators/Generator.inl#L42
https://github.com/Auburn/FastNoise2/blob/v0.9.5-alpha/include/FastNoise/Generators/Fractal.inl#L24
https://github.com/Auburn/FastNoise2/blob/v0.9.5-alpha/include/FastNoise/Generators/Fractal.inl#L15
https://github.com/Auburn/FastNoise2/blob/v0.9.5-alpha/include/FastNoise/Generators/Generator.inl#L102
```cpp
    fnFractal->GenUniformGrid2D(
            output, xOffset, yOffset, xSize, ySize, frequency, seed
    );
```

**Comment by jihiggins** [comment]
(and then no problems next run, etc)

**Comment by jihiggins** [comment]
Yep, this latest set of stuff is with that set to true. I'll add an assert for the fnSimplex check

**Comment by jihiggins** [comment]
(Sorry in advance if this turns out to be some wonky FFI thing haha)

**Comment by BasBuur** [comment]
I've ran into the exact same problem. I was able to partially track it down, for some reason `dynamic_cast<VoidPtrStorageType>( base );` returns null in `SetSourceSIMDPtr` in `Generator.inl`. That's the reason of the crash. Unfortunately in my project I cannot debug it further as the problem does not occur in debug builds (likely due to the fact that debug builds run a *lot* slower) and the variables I'd need for debugging it get optimized out somehow (or at least don't load on breakpoints).

Either way for now I'm simply re-creating the base generator and re-setting the source in derived generators when `simdGeneratorPtr` is null. Works, but obviously far from ideal.
